spot-ml/src/main/scala/org/apache/spot/SpotSparkLDAWrapper.scala

```diff
@@ -34,43 +35,59 @@ object SpotSparkLDAWrapper {
              ldaBeta: Double = 1.001,
              maxIterations: Int = 20): SpotLDAOutput = {
 
+    import sqlContext.implicits._
+
+    val docWordCountCache = docWordCount.cache()
+
+    // Forcing an action to cache results.
+    docWordCountCache.count()
+
     // Create word Map Word,Index for further usage
     val wordDictionary: Map[String, Int] = {
-      val words = docWordCount
-        .cache
+      val words = docWordCountCache
         .map({ case SpotLDAInput(doc, word, count) => word })
         .distinct
         .collect
       words.zipWithIndex.toMap
     }
 
-    val distinctDocument: Array[String] = docWordCount.map({ case SpotLDAInput(doc, word, count) => doc }).distinct.collect
+    val documentDictionary: DataFrame = docWordCountCache
+      .map({ case SpotLDAInput(doc, word, count) => doc })
+      .distinct
+      .zipWithIndex
+      .toDF(DocumentName, DocumentId)
+      .cache
 
-    // Create document Map Index, Document for further usage
-    val documentDictionary: Map[Int, String] = {
-      distinctDocument
-        .zipWithIndex
-        .sortBy(_._2)
-        .map(kv => (kv._2, kv._1))
-        .toMap
-    }
-    val docStrToID: Map[String, Int] = documentDictionary.map(_.swap)
+    // Forcing an action to cache results
+    documentDictionary.count()

```

Comment:
This is a different one. documentDictionary is just a data frame with IP and ID while docWordCount is used for word dictionary, model creation and document dictionary itself. 

As soon as they are not needed I'm explicitly unpersisting them (docWordCount and documentDictionary).